File: swift/Sources/Wildling.swift
final class Wildling {
    static let version = "1.0.0"

    private let generators: [Generator]
    private let patternCount: Int
    private var internalIndex: Int = 0

    init(patterns: [String], dictionaries: Dictionaries? = nil) {
        let dicts = dictionaries ?? [:]
        var gens: [Generator] = []
        var total = 0
        for pattern in patterns {
            let generator = Generator(pattern, dictionaries: dicts)
            gens.append(generator)
            total += generator.count()
        }
        self.generators = gens
        self.patternCount = total
    }

    func index() -> Int { internalIndex }

    func count() -> Int { patternCount }

    func reset() {
        internalIndex = 0
    }

    /// Next combination, or `nil` when exhausted.
    func next() -> String? {
        if internalIndex == patternCount {
            return nil
        }
        internalIndex += 1
        return get(internalIndex - 1)
    }

    func generatorsList() -> [Generator] { generators }

    /// Combination at index, or `nil` if out of range.
    func get(_ index: Int) -> String? {
        if index > patternCount - 1 || index < 0 {
            return nil
        }
        var segmentIndex = 0
        for generator in generators {
            let patternIndex = index - segmentIndex
            if patternIndex < generator.count() {
                return generator.get(patternIndex)
            }
            segmentIndex += generator.count()
        }
        return nil
    }
}

File: visualbasic/Wildling.vb
Public NotInheritable Class Wildling
    Public Const Version As String = "2.0.3"

    Private ReadOnly _generators As List(Of Generator)
    Private ReadOnly _patternCount As Integer
    Private _internalIndex As Integer

    Public Sub New(patterns As IEnumerable(Of String), Optional dictionaries As IReadOnlyDictionary(Of String, List(Of String)) = Nothing)
        Dim dicts = If(dictionaries, New Dictionary(Of String, List(Of String))())
        _generators = New List(Of Generator)()
        Dim total = 0
        For Each pattern In patterns
            Dim generator As New Generator(pattern, dicts)
            _generators.Add(generator)
            total += generator.Count()
        Next
        _patternCount = total
        _internalIndex = 0
    End Sub

    Public Shared Function Create(
        patterns As IEnumerable(Of String),
        Optional dictionaries As IReadOnlyDictionary(Of String, List(Of String)) = Nothing) As Wildling
        Return New Wildling(patterns, dictionaries)
    End Function

    Public Function Index() As Integer
        Return _internalIndex
    End Function

    Public Function Count() As Integer
        Return _patternCount
    End Function

    Public Sub Reset()
        _internalIndex = 0
    End Sub

    ''' <summary>Next combination, or <c>false</c> when exhausted.</summary>
    Public Function [Next]() As Object
        If _internalIndex = _patternCount Then
            Return False
        End If
        _internalIndex += 1
        Return [Get](_internalIndex - 1)
    End Function

    Public Function Generators() As IReadOnlyList(Of Generator)
        Return _generators
    End Function

    ''' <summary>Combination at index, or <c>false</c> if out of range.</summary>
    Public Function [Get](index As Integer) As Object
        If index > _patternCount - 1 OrElse index < 0 Then
            Return False
        End If

        Dim segmentIndex = 0
        For Each generator In _generators
            Dim patternIndex = index - segmentIndex
            If patternIndex < generator.Count() Then
                Return generator.Get(patternIndex)
            End If
            segmentIndex += generator.Count()
        Next
        Return False
    End Function
End Class

